lib/improver/nbhood.py

```diff
@@ -55,107 +59,243 @@ class BasicNeighbourhoodProcessing(object):
-
+        lead_times : None or List
+            List of lead times or forecast periods, at which the radii
+            within radii_in_km are defined.
         unweighted_mode : boolean
             If True, use a circle with constant weighting.
             If False, use a circle for neighbourhood kernel with
             weighting decreasing with radius.
 
         """
-        self.radius_in_km = float(radius_in_km)
+        if isinstance(radii_in_km, list):
+            self.radii_in_km = map(float, radii_in_km)
+        else:
+            self.radii_in_km = float(radii_in_km)
+        self.lead_times = lead_times
+        if self.lead_times is not None:
+            if len(radii_in_km) != len(lead_times):
+                msg = ("There is a mismatch in the number of radii "
+                       "and the number of lead times. "
+                       "Unable to continue due to mismatch.")
+                raise ValueError(msg)
         self.unweighted_mode = bool(unweighted_mode)
 
     def __str__(self):
-        result = ('<NeighbourhoodProcessing: radius_in_km: {};' +
+        result = ('<NeighbourhoodProcessing: radii_in_km: {};' +
                   'unweighted_mode: {}>')
         return result.format(
-            self.radius_in_km, self.unweighted_mode)
+            self.radii_in_km, self.unweighted_mode)
+
+    @staticmethod
+    def _find_required_lead_times(cube):
+        """
+        Determine the lead times within a cube, either by reading the
+        forecast_period coordinate, or by calculating the difference between
+        the time and the forecast_reference_time.
+
+        Parameters
+        ----------
+        cube : Iris.cube.Cube
+            Cube from which the lead times will be determined.
+
+        Returns
+        -------
+        required_lead_times : Numpy array
+            Array containing the lead times, at which the radii need to be
+            calculated.
+
+        """
+        if cube.coords("forecast_period"):
+            required_lead_times = cube.coord("forecast_period").points
+        else:
+            if cube.coords("time") and cube.coords("forecast_reference_time"):
+                required_lead_times = (
+                    cube.coord("time").points -
+                    cube.coord("forecast_reference_time").points)
+            else:
+                msg = ("The forecast period coordinate is not available "
+                       "within {}."
+                       "The time coordinate and forecast_reference_time "
+                       "coordinate were also not available for calculating "
+                       "the forecast_period.".format(cube))
+                raise CoordinateNotFoundError(msg)
+        return required_lead_times
 
-    def get_grid_x_y_kernel_ranges(self, cube):
-        """Return grid cell numbers east and north for the kernel."""
+    def _get_grid_x_y_kernel_ranges(self, cube, radii_in_km):
```

Comment:
As in process (below), it might be clearer here if this were radius_in_km as you are treating a single radius with each call to this function.